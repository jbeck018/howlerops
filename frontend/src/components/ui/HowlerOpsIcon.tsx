import React from 'react';

interface HowlerOpsIconProps {
  size?: number;
  className?: string;
  variant?: 'logo' | 'icon' | 'light' | 'dark';
}

export const HowlerOpsIcon: React.FC<HowlerOpsIconProps> = ({ 
  size = 32, 
  className = '',
  variant = 'icon'
}) => {
  // Use PNG icons for better quality and consistency
  const getIconSrc = () => {
    switch (variant) {
      case 'light':
        return '/src/assets/howlerops-icon-light.png';
      case 'dark':
        return '/src/assets/howlerops-icon-dark.png';
      case 'logo':
        return '/src/assets/howlerops-icon.png';
      default:
        return '/src/assets/howlerops-icon.png';
    }
  };

  return (
    <img 
      src={getIconSrc()}
      width={size} 
      height={size} 
      alt="HowlerOps"
      className={className}
    />
  );
};

export default HowlerOpsIcon;