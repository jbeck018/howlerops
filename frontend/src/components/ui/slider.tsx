import React from 'react'

type SliderProps = {
  value?: number[]
  defaultValue?: number[]
  max?: number
  min?: number
  step?: number
  onValueChange?: (value: number[]) => void
  className?: string
}

export function Slider({ value, defaultValue, max = 100, min = 0, step = 1, onValueChange, className }: SliderProps) {
  const [internal, setInternal] = React.useState<number>(defaultValue?.[0] ?? value?.[0] ?? 0)

  React.useEffect(() => {
    if (typeof value?.[0] === 'number') setInternal(value![0])
  }, [value?.[0]])

  return (
    <input
      type="range"
      className={className}
      max={max}
      min={min}
      step={step}
      value={internal}
      onChange={(e) => {
        const v = Number(e.target.value)
        setInternal(v)
        onValueChange?.([v])
      }}
    />
  )
}
