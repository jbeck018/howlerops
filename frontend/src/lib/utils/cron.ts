/**
 * Cron Expression Utilities
 * Parse and generate human-readable cron expressions
 */

export interface CronSchedule {
  minute: string
  hour: string
  dayOfMonth: string
  month: string
  dayOfWeek: string
}

export interface CronPreset {
  label: string
  description: string
  expression: string
}

export const CRON_PRESETS: CronPreset[] = [
  {
    label: 'Every Hour',
    description: 'Runs at the start of every hour',
    expression: '0 * * * *',
  },
  {
    label: 'Every 6 Hours',
    description: 'Runs every 6 hours starting at midnight',
    expression: '0 */6 * * *',
  },
  {
    label: 'Daily at 9 AM',
    description: 'Runs once per day at 9:00 AM',
    expression: '0 9 * * *',
  },
  {
    label: 'Daily at Midnight',
    description: 'Runs once per day at 12:00 AM',
    expression: '0 0 * * *',
  },
  {
    label: 'Weekly on Monday',
    description: 'Runs every Monday at 9:00 AM',
    expression: '0 9 * * 1',
  },
  {
    label: 'Monthly on 1st',
    description: 'Runs on the 1st of every month at 9:00 AM',
    expression: '0 9 1 * *',
  },
  {
    label: 'Weekdays at 9 AM',
    description: 'Runs Monday through Friday at 9:00 AM',
    expression: '0 9 * * 1-5',
  },
]

/**
 * Parse a cron expression into its components
 */
export function parseCronExpression(expression: string): CronSchedule {
  const parts = expression.split(' ')

  if (parts.length !== 5) {
    throw new Error('Invalid cron expression. Expected 5 parts.')
  }

  return {
    minute: parts[0],
    hour: parts[1],
    dayOfMonth: parts[2],
    month: parts[3],
    dayOfWeek: parts[4],
  }
}

/**
 * Build a cron expression from its components
 */
export function buildCronExpression(schedule: CronSchedule): string {
  return [
    schedule.minute,
    schedule.hour,
    schedule.dayOfMonth,
    schedule.month,
    schedule.dayOfWeek,
  ].join(' ')
}

/**
 * Convert cron expression to human-readable description
 */
export function cronToHumanReadable(expression: string): string {
  try {
    const cron = parseCronExpression(expression)

    // Check for presets
    const preset = CRON_PRESETS.find((p) => p.expression === expression)
    if (preset) {
      return preset.description
    }

    // Build description
    const parts: string[] = []

    // Frequency
    if (cron.minute === '*' && cron.hour === '*') {
      parts.push('Runs every minute')
    } else if (cron.minute.startsWith('*/')) {
      const interval = cron.minute.slice(2)
      parts.push(`Runs every ${interval} minutes`)
    } else if (cron.hour.startsWith('*/')) {
      const interval = cron.hour.slice(2)
      parts.push(`Runs every ${interval} hours`)
    } else if (cron.dayOfMonth !== '*' && cron.month !== '*') {
      parts.push(`Runs on day ${cron.dayOfMonth} of month ${cron.month}`)
    } else if (cron.dayOfWeek !== '*') {
      const day = getDayName(cron.dayOfWeek)
      parts.push(`Runs every ${day}`)
    } else if (cron.dayOfMonth !== '*') {
      parts.push(`Runs on day ${cron.dayOfMonth} of every month`)
    } else {
      parts.push('Runs daily')
    }

    // Time
    if (cron.hour !== '*' && cron.minute !== '*') {
      const time = formatTime(cron.hour, cron.minute)
      parts.push(`at ${time}`)
    }

    return parts.join(' ')
  } catch (error) {
    return 'Invalid cron expression'
  }
}

/**
 * Get next scheduled run time
 */
export function getNextRunTime(expression: string, from: Date = new Date()): Date {
  // This is a simplified version - in production use a library like cron-parser
  const cron = parseCronExpression(expression)
  const next = new Date(from)

  // Simple logic for common cases
  if (cron.minute !== '*') {
    next.setMinutes(parseInt(cron.minute, 10))
  }
  if (cron.hour !== '*') {
    next.setHours(parseInt(cron.hour, 10))
  }

  // If time has passed today, move to tomorrow
  if (next <= from) {
    next.setDate(next.getDate() + 1)
  }

  return next
}

/**
 * Validate cron expression
 */
export function isValidCronExpression(expression: string): boolean {
  try {
    const parts = expression.split(' ')
    if (parts.length !== 5) return false

    const validators = [
      (v: string) => isValidCronValue(v, 0, 59), // minute
      (v: string) => isValidCronValue(v, 0, 23), // hour
      (v: string) => isValidCronValue(v, 1, 31), // day of month
      (v: string) => isValidCronValue(v, 1, 12), // month
      (v: string) => isValidCronValue(v, 0, 6), // day of week
    ]

    return parts.every((part, i) => validators[i](part))
  } catch {
    return false
  }
}

function isValidCronValue(value: string, min: number, max: number): boolean {
  // Allow wildcards
  if (value === '*') return true

  // Allow ranges (e.g., 1-5)
  if (value.includes('-')) {
    const [start, end] = value.split('-').map((v) => parseInt(v, 10))
    return start >= min && end <= max && start <= end
  }

  // Allow intervals (e.g., */5)
  if (value.startsWith('*/')) {
    const interval = parseInt(value.slice(2), 10)
    return interval > 0 && interval <= max
  }

  // Allow lists (e.g., 1,3,5)
  if (value.includes(',')) {
    return value.split(',').every((v) => {
      const num = parseInt(v, 10)
      return num >= min && num <= max
    })
  }

  // Single value
  const num = parseInt(value, 10)
  return num >= min && num <= max
}

function getDayName(dayOfWeek: string): string {
  const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']

  if (dayOfWeek === '*') return 'day'

  const num = parseInt(dayOfWeek, 10)
  if (num >= 0 && num <= 6) {
    return days[num]
  }

  return 'day'
}

function formatTime(hour: string, minute: string): string {
  const h = parseInt(hour, 10)
  const m = parseInt(minute, 10)

  const period = h >= 12 ? 'PM' : 'AM'
  const displayHour = h === 0 ? 12 : h > 12 ? h - 12 : h
  const displayMinute = m.toString().padStart(2, '0')

  return `${displayHour}:${displayMinute} ${period}`
}

/**
 * Get a friendly relative time description for next run
 */
export function getRelativeNextRun(nextRunAt: string | Date): string {
  const next = typeof nextRunAt === 'string' ? new Date(nextRunAt) : nextRunAt
  const now = new Date()
  const diffMs = next.getTime() - now.getTime()

  if (diffMs < 0) return 'Overdue'

  const minutes = Math.floor(diffMs / 60000)
  const hours = Math.floor(minutes / 60)
  const days = Math.floor(hours / 24)

  if (days > 0) return `in ${days} day${days > 1 ? 's' : ''}`
  if (hours > 0) return `in ${hours} hour${hours > 1 ? 's' : ''}`
  if (minutes > 0) return `in ${minutes} minute${minutes > 1 ? 's' : ''}`
  return 'in less than a minute'
}
