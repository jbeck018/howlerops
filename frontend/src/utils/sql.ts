interface SqlStatementSegment {
  text: string
  start: number
  end: number
}

const BLOCK_COMMENT_REGEX = /\/\*[\s\S]*?\*\//g

const COMMENT_LINE_PREFIXES = ['--', '#']

const matchDollarTag = (source: string, start: number): string | null => {
  const remaining = source.slice(start)
  const match = /^\$[A-Za-z0-9_]*\$/.exec(remaining)
  return match ? match[0] : null
}

const pushStatement = (source: string, statements: SqlStatementSegment[], segmentStart: number, segmentEnd: number) => {
  if (segmentEnd <= segmentStart) {
    return
  }

  const rawSegment = source.slice(segmentStart, segmentEnd)
  const trimmed = rawSegment.trim()
  if (!trimmed) {
    return
  }

  const leadingWhitespaceLength = rawSegment.length - rawSegment.trimStart().length
  const trailingWhitespaceLength = rawSegment.length - rawSegment.trimEnd().length

  statements.push({
    text: trimmed,
    start: segmentStart + leadingWhitespaceLength,
    end: segmentEnd - trailingWhitespaceLength,
  })
}

const sanitizeStatementText = (statement: string): string => {
  if (!statement) {
    return ''
  }

  // Remove block comments
  let sanitized = statement.replace(BLOCK_COMMENT_REGEX, ' ')

  // Remove comment-only lines and trim trailing comment markers within a line
  const lines = sanitized
    .split('\n')
    .map((line) => {
      const trimmed = line.trim()
      if (!trimmed) {
        return ''
      }

      if (COMMENT_LINE_PREFIXES.some(prefix => trimmed.startsWith(prefix))) {
        return ''
      }

      return line
    })
    .filter(Boolean)

  sanitized = lines.join('\n')

  return sanitized.trim()
}

const parseStatements = (source: string): SqlStatementSegment[] => {
  const statements: SqlStatementSegment[] = []
  const length = source.length

  let index = 0
  let segmentStart = 0

  let inSingleQuote = false
  let inDoubleQuote = false
  let inDollarTag: string | null = null
  let inLineComment = false
  let inBlockComment = false

  while (index < length) {
    const char = source[index]
    const nextChar = index + 1 < length ? source[index + 1] : ''

    if (inLineComment) {
      if (char === '\n' || char === '\r') {
        inLineComment = false
      }
      index += 1
      continue
    }

    if (inBlockComment) {
      if (char === '*' && nextChar === '/') {
        inBlockComment = false
        index += 2
      } else {
        index += 1
      }
      continue
    }

    if (inDollarTag) {
      if (source.startsWith(inDollarTag, index)) {
        index += inDollarTag.length
        inDollarTag = null
      } else {
        index += 1
      }
      continue
    }

    if (inSingleQuote) {
      if (char === "'") {
        if (nextChar === "'") {
          index += 2
          continue
        }
        inSingleQuote = false
      }
      index += 1
      continue
    }

    if (inDoubleQuote) {
      if (char === '"') {
        if (nextChar === '"') {
          index += 2
          continue
        }
        inDoubleQuote = false
      }
      index += 1
      continue
    }

    if (char === '-' && nextChar === '-') {
      inLineComment = true
      index += 2
      continue
    }

    if (char === '/' && nextChar === '*') {
      inBlockComment = true
      index += 2
      continue
    }

    if (char === "'") {
      inSingleQuote = true
      index += 1
      continue
    }

    if (char === '"') {
      inDoubleQuote = true
      index += 1
      continue
    }

    if (char === '$') {
      const tag = matchDollarTag(source, index)
      if (tag) {
        inDollarTag = tag
        index += tag.length
        continue
      }
    }

    if (char === ';') {
      pushStatement(source, statements, segmentStart, index)
      segmentStart = index + 1
    }

    index += 1
  }

  pushStatement(source, statements, segmentStart, length)

  return statements
}

const findStatementIndexForCursor = (statements: SqlStatementSegment[], cursor: number): number => {
  if (!statements.length) {
    return -1
  }

  const strictMatch = statements.findIndex((statement) => cursor >= statement.start && cursor <= statement.end)
  if (strictMatch !== -1) {
    return strictMatch
  }

  const nextMatch = statements.findIndex((statement) => cursor <= statement.start)
  if (nextMatch !== -1) {
    return nextMatch
  }

  return statements.length - 1
}

const getSanitizedStatementAtIndex = (statements: SqlStatementSegment[], index: number): string | null => {
  if (index < 0 || index >= statements.length) {
    return null
  }

  const cleaned = sanitizeStatementText(statements[index].text)
  if (cleaned) {
    return cleaned
  }

  // If the targeted statement was effectively empty after sanitisation, try to find the next meaningful statement
  for (let i = index + 1; i < statements.length; i += 1) {
    const nextCleaned = sanitizeStatementText(statements[i].text)
    if (nextCleaned) {
      return nextCleaned
    }
  }

  for (let i = index - 1; i >= 0; i -= 1) {
    const previousCleaned = sanitizeStatementText(statements[i].text)
    if (previousCleaned) {
      return previousCleaned
    }
  }

  return null
}

export const buildExecutableSql = (
  documentText: string,
  options: {
    selectionText?: string
    cursorOffset: number
  }
): string | null => {
  const { selectionText, cursorOffset } = options

  if (selectionText) {
    const cleanedSelection = sanitizeStatementText(selectionText)
    if (cleanedSelection) {
      return cleanedSelection
    }
  }

  const statements = parseStatements(documentText)
  if (!statements.length) {
    return null
  }

  const targetIndex = findStatementIndexForCursor(statements, cursorOffset)
  return getSanitizedStatementAtIndex(statements, targetIndex)
}

export const collapseWhitespace = (sql: string): string =>
  sql
    .split('\n')
    .map(line => line.trim())
    .filter(Boolean)
    .join('\n')
