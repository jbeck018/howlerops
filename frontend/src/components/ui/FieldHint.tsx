import * as React from "react"
import { cn } from "@/lib/utils"
import { AlertCircle, Info } from "lucide-react"

export interface FieldHintProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode
  variant?: "default" | "error" | "warning"
  icon?: boolean
}

export function FieldHint({
  children,
  variant = "default",
  icon = true,
  className,
  ...props
}: FieldHintProps) {
  const Icon = variant === "error" || variant === "warning" ? AlertCircle : Info

  return (
    <div
      className={cn(
        "flex items-start gap-2 text-sm mt-1.5",
        variant === "default" && "text-muted-foreground",
        variant === "error" && "text-destructive",
        variant === "warning" && "text-amber-600 dark:text-amber-500",
        className
      )}
      {...props}
    >
      {icon && <Icon className="h-4 w-4 mt-0.5 flex-shrink-0" />}
      <div className="flex-1">{children}</div>
    </div>
  )
}
