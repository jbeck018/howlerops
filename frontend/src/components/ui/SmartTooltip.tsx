import * as React from "react"
import * as TooltipPrimitive from "@radix-ui/react-tooltip"
import { cn } from "@/lib/utils"

export interface SmartTooltipProps {
  children: React.ReactNode
  content: React.ReactNode
  side?: "top" | "right" | "bottom" | "left"
  align?: "start" | "center" | "end"
  delayDuration?: number
  disabled?: boolean
  keyboardShortcut?: string
  className?: string
}

export function SmartTooltip({
  children,
  content,
  side = "top",
  align = "center",
  delayDuration = 500,
  disabled = false,
  keyboardShortcut,
  className,
}: SmartTooltipProps) {
  if (disabled) {
    return <>{children}</>
  }

  return (
    <TooltipPrimitive.Provider delayDuration={delayDuration}>
      <TooltipPrimitive.Root>
        <TooltipPrimitive.Trigger asChild>
          {children}
        </TooltipPrimitive.Trigger>
        <TooltipPrimitive.Portal>
          <TooltipPrimitive.Content
            side={side}
            align={align}
            className={cn(
              "z-50 overflow-hidden rounded-md border bg-popover px-3 py-1.5 text-sm text-popover-foreground shadow-md animate-in fade-in-0 zoom-in-95 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2",
              className
            )}
            sideOffset={5}
          >
            <div className="flex flex-col gap-1">
              <div>{content}</div>
              {keyboardShortcut && (
                <div className="flex items-center gap-1 text-xs text-muted-foreground">
                  <kbd className="pointer-events-none inline-flex h-5 select-none items-center gap-1 rounded border bg-muted px-1.5 font-mono text-[10px] font-medium opacity-100">
                    {keyboardShortcut}
                  </kbd>
                </div>
              )}
            </div>
            <TooltipPrimitive.Arrow className="fill-popover" />
          </TooltipPrimitive.Content>
        </TooltipPrimitive.Portal>
      </TooltipPrimitive.Root>
    </TooltipPrimitive.Provider>
  )
}
