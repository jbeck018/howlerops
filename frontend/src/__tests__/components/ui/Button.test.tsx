import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, userEvent } from '../../utils/test-utils';
import { Button } from '@/components/ui/button';

describe('Button Component', () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    user = userEvent.setup();
  });
  it('renders with default props', () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole('button', { name: /click me/i });
    expect(button).toBeInTheDocument();
    expect(button).toHaveClass('bg-primary');
  });

  it('renders with different variants', () => {
    const { rerender } = render(<Button variant="default">Default</Button>);
    expect(screen.getByRole('button')).toHaveClass('bg-primary');

    rerender(<Button variant="destructive">Destructive</Button>);
    expect(screen.getByRole('button')).toHaveClass('bg-destructive');

    rerender(<Button variant="outline">Outline</Button>);
    expect(screen.getByRole('button')).toHaveClass('border-input');

    rerender(<Button variant="secondary">Secondary</Button>);
    expect(screen.getByRole('button')).toHaveClass('bg-secondary');

    rerender(<Button variant="ghost">Ghost</Button>);
    expect(screen.getByRole('button')).toHaveClass('hover:bg-accent');

    rerender(<Button variant="link">Link</Button>);
    expect(screen.getByRole('button')).toHaveClass('text-primary');
  });

  it('renders with different sizes', () => {
    const { rerender } = render(<Button size="default">Default</Button>);
    expect(screen.getByRole('button')).toHaveClass('h-10');

    rerender(<Button size="sm">Small</Button>);
    expect(screen.getByRole('button')).toHaveClass('h-9');

    rerender(<Button size="lg">Large</Button>);
    expect(screen.getByRole('button')).toHaveClass('h-11');

    rerender(<Button size="icon">Icon</Button>);
    expect(screen.getByRole('button')).toHaveClass('h-10', 'w-10');
  });

  it('handles click events', async () => {
    const handleClick = vi.fn();

    render(<Button onClick={handleClick}>Click me</Button>);
    const button = screen.getByRole('button');

    await user.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('can be disabled', async () => {
    const handleClick = vi.fn();

    render(<Button disabled onClick={handleClick}>Disabled</Button>);
    const button = screen.getByRole('button');

    expect(button).toBeDisabled();
    await user.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });

  it('supports custom className', () => {
    render(<Button className="custom-class">Custom</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveClass('custom-class');
  });

  it('forwards ref correctly', () => {
    let buttonRef: HTMLButtonElement | null = null;
    const RefComponent = () => (
      <Button ref={(ref) => { buttonRef = ref; }}>Ref Button</Button>
    );

    render(<RefComponent />);
    expect(buttonRef).toBeInstanceOf(HTMLButtonElement);
    expect(buttonRef?.textContent).toBe('Ref Button');
  });

  it('renders as different element when asChild is true', () => {
    render(
      <Button asChild>
        <a href="/test">Link Button</a>
      </Button>
    );

    const link = screen.getByRole('link');
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/test');
    expect(link).toHaveClass('bg-primary'); // Should still have button styles
  });

  it('supports keyboard navigation', async () => {
    const handleClick = vi.fn();

    render(<Button onClick={handleClick}>Keyboard Button</Button>);
    const button = screen.getByRole('button');

    button.focus();
    expect(button).toHaveFocus();

    await user.keyboard('{Enter}');
    expect(handleClick).toHaveBeenCalledTimes(1);

    await user.keyboard(' ');
    expect(handleClick).toHaveBeenCalledTimes(2);
  });

  it('has proper accessibility attributes', () => {
    render(<Button aria-label="Accessible button">Icon only</Button>);
    const button = screen.getByRole('button');
    expect(button).toHaveAttribute('aria-label', 'Accessible button');
  });

  it('supports loading state', () => {
    render(<Button disabled>Loading...</Button>);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    expect(button).toHaveTextContent('Loading...');
  });

  describe('Button variants visual regression', () => {
    it('matches snapshot for all variants', () => {
      const variants = ['default', 'destructive', 'outline', 'secondary', 'ghost', 'link'] as const;

      variants.forEach(variant => {
        const { container } = render(<Button variant={variant}>{variant}</Button>);
        expect(container.firstChild).toMatchSnapshot(`button-${variant}`);
      });
    });

    it('matches snapshot for all sizes', () => {
      const sizes = ['default', 'sm', 'lg', 'icon'] as const;

      sizes.forEach(size => {
        const { container } = render(<Button size={size}>{size}</Button>);
        expect(container.firstChild).toMatchSnapshot(`button-${size}`);
      });
    });
  });

  describe('Button interaction states', () => {
    it('shows hover state on mouse enter', async () => {
      render(<Button>Hover me</Button>);
      const button = screen.getByRole('button');

      await user.hover(button);
      expect(button).toHaveClass('hover:bg-primary/90');
    });

    it('shows focus state on keyboard focus', async () => {
      render(<Button>Focus me</Button>);
      const button = screen.getByRole('button');

      button.focus();
      expect(button).toHaveFocus();
      expect(button).toHaveClass('focus-visible:ring-2');
    });

    it('supports tab navigation', async () => {
      render(
        <div>
          <Button>First</Button>
          <Button>Second</Button>
          <Button disabled>Disabled</Button>
          <Button>Third</Button>
        </div>
      );

      const firstButton = screen.getByRole('button', { name: 'First' });
      const secondButton = screen.getByRole('button', { name: 'Second' });
      const thirdButton = screen.getByRole('button', { name: 'Third' });

      firstButton.focus();
      expect(firstButton).toHaveFocus();

      await user.tab();
      expect(secondButton).toHaveFocus();

      await user.tab();
      expect(thirdButton).toHaveFocus(); // Should skip disabled button
    });
  });
});