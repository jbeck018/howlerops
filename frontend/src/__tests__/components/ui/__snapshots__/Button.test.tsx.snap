// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Button variants visual regression > matches snapshot for all sizes > button-default 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
>
  default
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all sizes > button-icon 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 w-10"
>
  icon
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all sizes > button-lg 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-11 rounded-md px-8"
>
  lg
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all sizes > button-sm 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-9 rounded-md px-3"
>
  sm
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all variants > button-default 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
>
  default
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all variants > button-destructive 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-destructive text-destructive-foreground hover:bg-destructive/90 h-10 px-4 py-2"
>
  destructive
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all variants > button-ghost 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
>
  ghost
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all variants > button-link 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 text-primary underline-offset-4 hover:underline h-10 px-4 py-2"
>
  link
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all variants > button-outline 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
>
  outline
</button>
`;

exports[`Button Component > Button variants visual regression > matches snapshot for all variants > button-secondary 1`] = `
<button
  class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 px-4 py-2"
>
  secondary
</button>
`;
